### PR TITLE
[10.0][ADD] sale_double_validation

### PR DESCRIPTION
**Pull request by leemannd**
Porting https://github.com/OCA/sale-workflow/pull/423 from @yvaucher in V10.0

**Double validation for sale**

This module adds an option to enable two steps validation.

When a minimal amount is reached on a Sale order, if a salesman confirm the sale.
The sale order will be set to "To approve" state and will require a Sales manager
to Approve it.

This module is base on the purchase option for two levels validation in module purchase.

### CODE REVIEW

**Review by gurneyalex — APPROVED** [review]

File: sale_double_validation/README.rst [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+==========================
+Double validation for sale
+==========================
+
+This module adds an option to enable two steps validation.
+
+When a minimal amount is reached on a Sale order, if a salesman confirm the sale.
```
Comment (guewen) [inline sale_double_validation/README.rst]: s/salesman/salesperson/

File: sale_double_validation/README.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+The sale order will be set to "To approve" state and will require a Sales manager
+to Approve it.
+
+This module is base on the purchase option for two levels validation in module `purchase`.
```
Comment (guewen) [inline sale_double_validation/README.rst]: s/base/based/

File: sale_double_validation/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+{'name': 'Double validation for Sales',
+ 'summary': "",
+ 'version': '10.0.1.0.0',
+ 'author': 'Camptocamp,Odoo Community Association (OCA)',
+ 'maintainer': 'Camptocamp',
+ 'license': 'AGPL-3',
+ 'category': 'category',
+ 'complexity': "complexity",
```
Comment (guewen) [inline sale_double_validation/__manifest__.py]: you can remove `complexity`

File: sale_double_validation/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright 2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+{'name': 'Double validation for Sales',
+ 'summary': "",
+ 'version': '10.0.1.0.0',
+ 'author': 'Camptocamp,Odoo Community Association (OCA)',
+ 'maintainer': 'Camptocamp',
+ 'license': 'AGPL-3',
+ 'category': 'category',
```
Comment (guewen) [inline sale_double_validation/__manifest__.py]: you should set something there

**Review by damdam-s — APPROVED** [review]

File: sale_double_validation/models/sale.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+# Copyright 2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    state = fields.Selection(
+        selection_add=[('to_approve', 'To Approve')]
```
Comment (guewen) [inline sale_double_validation/models/sale.py]: State will be added at the far right of the state widget which is not nice.

Instead, you can add the state at the right place with:

```python

    @api.model
    def _setup_fields(self, partial):
        super(SaleOrder, self)._setup_fields(partial)
        selection = self._fields['state'].selection
        position = 0
        exists = False
        for idx, (state, __) in enumerate(selection):
            if state == 'draft':
                position = idx
            elif state == 'to_approve':
                exists = True
        if not exists:
            selection.insert(position + 1, ('to_approve', _('To Approve')))

```

File: sale_double_validation/models/sale.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    )
+
+    @api.multi
+    def is_amount_to_approve(self):
+        currency = self.env.user.company_id.currency_id
+        limit_amount = self.company_id.so_double_validation_amount
+        limit_amount = currency.compute(limit_amount, self.currency_id)
+        return self.amount_total >= limit_amount
```
Comment (guewen) [inline sale_double_validation/models/sale.py]: Comparison of floats should preferably be done with `odoo.tools.float_compare()`.

File: sale_double_validation/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    @api.multi
+    def is_to_approve(self):
+        return (self.company_id.so_double_validation == 'two_step' and
+                self.is_amount_to_approve() and
+                not self.user_has_groups('sales_team.group_sale_manager'))
+
+    @api.multi
+    def action_confirm(self):
+        to_approve = self.env['sale.order']
+        to_confirm = self.env['sale.order']
```
Comment (guewen) [inline sale_double_validation/models/sale.py]: This is really confusing and looks like an error on fast reading of the code.
Even if under the cover it has the same effect, I suggest the following lines. I think they better show the intent which is to create new recordset containers for gathering records:

```
to_approve = self.env['sale.order'].browse()
to_confirm = self.env['sale.order'].browse()
```

**Review by guewen — CHANGES_REQUESTED** [review]
I know that most of the code is from #423 but could you take the opportunity to change a few things? thanks

I close #423

**Comment by leemannd** [comment]
@guewen Hum, why closing #423 It targets V9.0 ?

**Comment by guewen** [comment]
> @guewen Hum, why closing #423 It targets V9.0 ?

Fixed. Misunderstanding.

**Comment by leemannd** [comment]
@guewen Thanks for the comments. Updated

File: sale_double_validation/models/sale.py [hunk 8]
```diff
@@ -1,22 +1,37 @@
...
+            selection.insert(position + 1, ('to_approve', _('To Approve')))
 
     @api.multi
     def is_amount_to_approve(self):
         currency = self.env.user.company_id.currency_id
         limit_amount = self.company_id.so_double_validation_amount
         limit_amount = currency.compute(limit_amount, self.currency_id)
-        return self.amount_total >= limit_amount
+        if float_compare(
+           self.amount_total, limit_amount, precision_digits=2) == 1:
```
Comment (guewen) [inline sale_double_validation/models/sale.py]: You can't hardcode the precision_digits which depends of the currency. 
Should be `precision_rounding=self.currency_id.rounding`.

Another point: you could shorten by returning the boolean result `return (float_compare... == 1)`

Comment (guewen) [reply sale_double_validation/models/sale.py]: Not sure why my comment is shown as outdated...

File: sale_double_validation/models/sale.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        if not exists:
+            selection.insert(position + 1, ('to_approve', _('To Approve')))
+
+    @api.multi
+    def is_amount_to_approve(self):
+        currency = self.env.user.company_id.currency_id
+        limit_amount = self.company_id.so_double_validation_amount
+        limit_amount = currency.compute(limit_amount, self.currency_id)
+        if float_compare(
+           limit_amount, self.amount_total, precision_digits=2) <= 0:
```
Comment (guewen) [inline sale_double_validation/models/sale.py]: You can't hardcode the precision_digits which depends of the currency. 
Should be `precision_rounding=self.currency_id.rounding`.

Another point: you could shorten by returning the boolean result `return (float_compare... <= 0)`

**Review by guewen — CHANGES_REQUESTED** [review]

**Comment by yvaucher** [comment]
For your first commit you should have done a git cherry-pick

**Comment by rafaelbn** [comment]
Hi @leemannd we are reviewing in the OCA Sprint code, could you review last comment from @yvaucher  please?
@guewen you comment is blocking , could you make last review please? thanks!

**Comment by leemannd** [comment]
@rafaelbn Fixed history to respect author's work.
@guewen PR is fixed.

**Comment by grindtildeath** [comment]
@leemannd Please merge https://github.com/leemannd/sale-workflow/pull/8

**Review by damdam-s — APPROVED** [review]

File: sale_double_validation/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+ 'version': '10.0.1.0.0',
+ 'author': 'Camptocamp,Odoo Community Association (OCA)',
+ 'maintainer': 'Camptocamp',
+ 'license': 'AGPL-3',
+ 'category': 'sale',
+ 'depends': [
+     'sale',
+     'sales_team',
+ ],
+ 'website': 'www.camptocamp.com',
```
Comment (nikul-serpentcs) [inline sale_double_validation/__manifest__.py]: @leemannd IMO change URL as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L10)

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

File: sale_double_validation/__manifest__.py (hunk 10 above)
Comment (leemannd) [reply sale_double_validation/__manifest__.py]: @nikul-serpentcs You're right. Rules have changed since the first commit of the PR. Updating it

**Review by mpanarin — APPROVED** [review]

**Comment by leemannd** [comment]
@nikul-serpentcs I've added a fixup. Can you update your review?

**Comment by pedrobaeza** [comment]
Can you please check Travis status?

File: sale_double_validation/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+ 'license': 'AGPL-3',
+ 'category': 'sale',
+ 'depends': [
+     'sale',
+     'sales_team',
+ ],
+ 'website': 'https://github.com/OCA/sale-workflow',
+ 'data': ['views/res_config.xml',
+          'views/sale.xml'],
+ 'test': [],
```
Comment (nikul-serpentcs) [inline sale_double_validation/__manifest__.py:19]: @leemannd Remove, No need here

**Review by nikul-serpentcs — COMMENTED** [review]
Small Change, Other Than LGTM

File: sale_double_validation/models/sale.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            if state == 'draft':
+                position = idx
+            elif state == 'to_approve':
+                exists = True
+        if not exists:
+            selection.insert(position + 1, ('to_approve', _('To Approve')))
+
+    @api.multi
+    def is_amount_to_approve(self):
+        currency = self.company_id.currency_id
```
Comment (SimoRubi) [inline sale_double_validation/models/sale.py:28]: `self.ensure_one()`

File: sale_double_validation/models/sale.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        limit_amount = self.company_id.so_double_validation_amount
+        limit_amount = currency.compute(limit_amount, self.currency_id)
+        rounding = self.currency_id.rounding
+        return float_compare(
+            limit_amount, self.amount_total, precision_rounding=rounding) <= 0
+
+    @api.multi
+    def is_to_approve(self):
+        return (self.company_id.so_double_validation == 'two_step' and
```
Comment (SimoRubi) [inline sale_double_validation/models/sale.py:38]: `self.ensure_one()`

File: sale_double_validation/tests/test_double_validation.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        so.company_id.sudo().so_double_validation = 'two_step'
+        # confirm quotation
+        so.sudo(self.user).action_confirm()
+        self.assertEquals(so.state, 'sale')
+
+    def test_two_steps_manager(self):
+        so = self.so
+        so.company_id.sudo().so_double_validation = 'two_step'
+        so.company_id.sudo().so_double_validation_amount = 10
+        # confirm quotation
```
Comment (SimoRubi) [inline sale_double_validation/tests/test_double_validation.py:28]: check that the total amount of the order is what expected (greater/less than `10`)

File: sale_double_validation/tests/test_double_validation.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        so.sudo(self.user).action_confirm()
+        self.assertEquals(so.state, 'to_approve')
+        so.sudo(self.manager).action_approve()
+        self.assertEquals(so.state, 'sale')
+
+    def test_two_steps_above_limit(self):
+        so = self.so
+        so.company_id.sudo().so_double_validation = 'two_step'
+        so.company_id.sudo().so_double_validation_amount = 10
```
Comment (SimoRubi) [inline sale_double_validation/tests/test_double_validation.py:45]: check that the total amount of the order is what expected (greater/less than `10`)

File: sale_double_validation/README.rst [hunk 16]
```diff
@@ -0,0 +1,80 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (SimoRubi) [inline sale_double_validation/README.rst:1]: Consider generating a nice README with https://github.com/OCA/maintainer-tools/blob/master/tools/gen_addon_readme.py

**Review by SimoRubi — CHANGES_REQUESTED** [review]
Just code review

File: sale_double_validation/models/sale.py (hunk 12 above)
Comment (nikul-serpentcs) [reply sale_double_validation/models/sale.py:28]: here use `multi ` so not need `self.ensure_one()` only add loop

**Comment by rafaelbn** [comment]
hi @leemannd could you please review travis :red_circle: ? thanks!

Please @guewen unblock here we should merge today! :smile:

**Comment by leemannd** [comment]
@rafaelbn @pedrobaeza  @SimoRubi  @nikul-serpentcs As it was done following rules when created, and as it's a 1,5 year old work and I didn't had the chance to merge it, feel free to make a PR to fix it or either close it.

File: sale_double_validation/README.rst (hunk 16 above)
Comment (leemannd) [reply sale_double_validation/README.rst:1]: Please consider merging it as the PR was made  on 2017/04/05.
I'm unwilling to change this.

**Comment by rafaelbn** [comment]
Hello @leemannd , sorry for this. But this PR is blocked by:

- @guewen from 2017
- Travis :red_circle: 

We were reviewing in CS and it would been merged if travis :green_apple:

**Comment by leemannd** [comment]
Taking a look

**Pull request closed without merging**